messaging/coap/observe.c

```diff
@@ -433,172 +435,196 @@ coap_remove_observer_by_resource(const oc_resource_t *rsc)
-                   "for client liveness");
-            notification->type = COAP_TYPE_CON;
-          }
-          coap_set_payload(
-            notification, response->response_buffer->buffer,
-            (uint32_t)response->response_buffer->response_length);
-        } //! blockwise transfer
-      }   // !is_revert
-
-      coap_set_status_code(notification, response->response_buffer->code);
-      if (notification->code < BAD_REQUEST_4_00 &&
-          obs->resource->num_observers) {
-        coap_set_header_observe(
-          notification, observe_increment_observe_counter(&obs->obs_counter));
-        observe_increment_observe_counter(&g_observe_counter);
-      } else {
-        coap_set_header_observe(notification,
-                                OC_COAP_OPTION_OBSERVE_UNREGISTER);
-      }
-      if (response->response_buffer->content_format > 0) {
-        coap_set_header_content_format(
-          notification, response->response_buffer->content_format);
+        OC_DBG("send_notification: forcing CON notification to check for "
+               "client liveness");
+        notification.type = COAP_TYPE_CON;
       }
-      coap_set_token(notification, obs->token, obs->token_len);
-      transaction = coap_new_transaction(coap_get_mid(), obs->token,
-                                         obs->token_len, &obs->endpoint);
-      if (transaction) {
-        obs->last_mid = transaction->mid;
-        notification->mid = transaction->mid;
-        transaction->message->length = coap_serialize_message(
-          notification, transaction->message->data, oc_message_buffer_size());
-        if (transaction->message->length > 0) {
-          coap_send_transaction(transaction);
-        } else {
-          coap_clear_transaction(transaction);
-        }
-      } // transaction
-    }   // response_buf != NULL
-  }     //! separate response
+      coap_set_payload(&notification, response->response_buffer->buffer,
+                       (uint32_t)response->response_buffer->response_length);
+    } //! blockwise transfer
+  }   // !is_revert
+
+  coap_set_status_code(&notification, response->response_buffer->code);
+  if (notification.code < BAD_REQUEST_4_00 && obs->resource->num_observers) {
+    coap_set_header_observe(
+      &notification, observe_increment_observe_counter(&obs->obs_counter));
+    observe_increment_observe_counter(&g_observe_counter);
+  } else {
+    coap_set_header_observe(&notification, OC_COAP_OPTION_OBSERVE_UNREGISTER);
+  }
+  if (response->response_buffer->content_format > 0) {
+    coap_set_header_content_format(&notification,
+                                   response->response_buffer->content_format);
+  }
+  coap_set_token(&notification, obs->token, obs->token_len);
+  transaction = coap_new_transaction(coap_get_mid(), obs->token, obs->token_len,
+                                     &obs->endpoint);
+  if (transaction == NULL) {
+    return 0;
```

Comment:
free notification is not needed ?

Reply:
seems to be a stack object, doesn't allocate anything, so free/clear is not necessary